### PR TITLE
improve console perfs

### PR DESCRIPTION
**Pull request by ydaveluy**
The console is slow when using unlimited console output with a lot of links.
This PR provides a faster implementation.

Use an internal list of positions that permit faster access, avoid
unnecessary copies and reduce memory consumption compared to IDocument
Position handling.

Here are some benchmarks for the execution of this code:
`

	public static void main(String[] args) {
	
		NumberFormatException e = new NumberFormatException("Invalid number: z");
		long start = System.nanoTime();
		for (int i = 0; i < 1_000_000; ++i) {
			e.printStackTrace(System.out);
		}
		System.out.println((System.nanoTime() - start) / 1_000_000_000.+"s");
	}

`

with unlimited console output:
- before: 49s
- after: 19s
--> almost impossible to scroll the console before this PR

### CODE REVIEW

File: org.eclipse.ui.console/src/org/eclipse/ui/console/TextConsole.java [hunk 1]
```diff
@@ -400,25 +395,24 @@ public IHyperlink getHyperlink(int offset) {
...
 	 */
-	private Position findPosition(int offset, Position[] positions) {
+	private ConsoleHyperlinkPosition findPosition(int offset) {
 
-		if (positions.length == 0) {
+		if (fPositions.isEmpty()) {
 			return null;
 		}
 
 		int left= 0;
```
Comment (Bananeweizen) [inline org.eclipse.ui.console/src/org/eclipse/ui/console/TextConsole.java:406]: Can't we replace all this selfmade searching in this method with a call to https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Arrays.html#binarySearch(T%5B%5D,int,int,T,java.util.Comparator) ?

File: org.eclipse.ui.console/src/org/eclipse/ui/console/TextConsole.java [hunk 2]
```diff
@@ -545,19 +539,39 @@ private synchronized void checkFinished() {
...
+		fPositions.add(getInsertionIndex(offset), hyperlinkPosition);
+		fConsoleManager.refresh(this);
+	}
+
+	private int getInsertionIndex(int offset) {
+		if (fPositions.isEmpty()) {
+			return 0;
 		}
+
+		var left = 0;
```
Comment (Bananeweizen) [inline org.eclipse.ui.console/src/org/eclipse/ui/console/TextConsole.java:558]: same here

Comment (ydaveluy) [reply org.eclipse.ui.console/src/org/eclipse/ui/console/TextConsole.java:558]: yes in this case it is possible !

Comment (ydaveluy) [reply org.eclipse.ui.console/src/org/eclipse/ui/console/TextConsole.java:558]: in fact it is not possible, the binarySearch returns -1 if the key is not found. An other option would be to let ConsoleHyperlinkPosition implement Comparable and store positions in a set.

File: org.eclipse.ui.console/src/org/eclipse/ui/console/TextConsole.java (hunk 1 above)
Comment (ydaveluy) [reply org.eclipse.ui.console/src/org/eclipse/ui/console/TextConsole.java]: Yes done !

File: org.eclipse.ui.console/src/org/eclipse/ui/console/TextConsole.java (hunk 2 above)
Comment (ydaveluy) [reply org.eclipse.ui.console/src/org/eclipse/ui/console/TextConsole.java]: done too !

**Comment by mihnita** [comment]
This affects any kind of positions.
It would be really great if it would be possible to fix the official document handling of positions instead of just use a different mechanism for hyperlinks.

Some of preliminary digging I did for https://github.com/mihnita/ansi-econsole/issues/66:
When the console buffer is "trimmed" it dumps a lot of lines from the beginning and shifts everything
The operation updates the positions, and in the process removes the ones that are in the deleted lines.
There are 2 arrays for positions, one for starts, sorted, and one for ends, unsorted.

But the operation is something like this:
```
for each position in allPositions
    if position needs to be deleted // in the line ranges to be deleted
        delete position document.getStartingPositions /*sorted*/ true 
        delete position document.getEndingPositions /*sorted*/ false
```
Note: these are not real names. I did the research, but didn't write them down.
I can do it if the direction seems promissing.

The start / end positions are stored in `ArrayList`
Each delete happens at the beginning, and the whole array is moved (`ArrayList` calls `System.arraycopy` in `remove`).
Do that for tens of thousands of deleted lines, and for several positions per line, and you have a performance problem.

---

Possible solution:

There are probably kinds of documents that always
- add positions at the end
- the positions never overlap
- delete lines from the beginning

Probably any doc attached to a console behaves like that.

The hyperlinks and ansi escapes are examples of non-overlapping positions.

For the solution it would be nice if we detect that and make the list of endings also sorted.

And then, of the document deletion happens from the beginning, find the range of positions to delete, and call `ArrayList.removeRange` instead of `remove`

**Comment by ydaveluy** [comment]
I would like to add some information: ending positions in AbstractDocument are sorted by end offset whereas starting positions are sorted by start offset. It is used by `AbstractDocument.getPositions(String, int, int, boolean, boolean)` to find overlapping positions for a given range.
As you said, in case of non overlapping positions such as hyperlinks and ansi escapes, it is possible to use a faster algo to find positions for a given range.

`ArrayList.removeRange'` is a protected method. It seems that the correct way is `ArrayList.subList(start, end).clear()`

File: org.eclipse.ui.console/src/org/eclipse/ui/console/TextConsole.java [hunk 3]
```diff
@@ -545,19 +539,39 @@ private synchronized void checkFinished() {
...
+		var right = fPositions.size() - 1;
+
+		int mid;
```
Comment (lbownik) [inline org.eclipse.ui.console/src/org/eclipse/ui/console/TextConsole.java]: how about instantiating 'mid' and 'position' within the loop?

File: org.eclipse.ui.console/src/org/eclipse/ui/console/TextConsoleViewer.java [hunk 4]
```diff
@@ -196,20 +196,30 @@ private boolean checkStartOfDocument() {
...
-						position.delete();
-					}
-					if (position.isDeleted) {
-						document.removePosition(ConsoleHyperlinkPosition.HYPER_LINK_CATEGORY, position);
-					}
+		// adapt existing positions (we are interested only by remove events)
+		if (event.getOffset() == 0) {
+
+			final var length = event.getLength();
+			var positions = console.getPositions();
```
Comment (lbownik) [inline org.eclipse.ui.console/src/org/eclipse/ui/console/TextConsoleViewer.java:206]: it seems that 'positions' variable is redundant

File: org.eclipse.ui.console/src/org/eclipse/ui/console/TextConsole.java [hunk 5]
```diff
@@ -362,17 +369,12 @@ public void clearConsole() {
...
-			for (int i = 0; i < positions.length; i++) {
-				ConsoleHyperlinkPosition position = (ConsoleHyperlinkPosition) positions[i];
-				hyperlinks[i] = position.getHyperLink();
-			}
-			return hyperlinks;
-		} catch (BadPositionCategoryException e) {
-			return new IHyperlink[0];
+		IHyperlink[] hyperlinks = new IHyperlink[fPositions.size()];
+		for (int i = 0; i < fPositions.size(); i++) {
+			ConsoleHyperlinkPosition position = fPositions.get(i);
```
Comment (lbownik) [inline org.eclipse.ui.console/src/org/eclipse/ui/console/TextConsole.java:375]: this variable seems redundant

File: org.eclipse.ui.console/src/org/eclipse/ui/console/TextConsole.java [hunk 6]
```diff
@@ -400,25 +395,24 @@ public IHyperlink getHyperlink(int offset) {
...
+		ConsoleHyperlinkPosition position = null;
 
 		while (left < right) {
 
 			mid= (left + right) / 2;
 
-			position= positions[mid];
+			position = fPositions.get(mid);
 			if (offset < position.getOffset()) {
 				if (left == mid) {
```
Comment (lbownik) [inline org.eclipse.ui.console/src/org/eclipse/ui/console/TextConsole.java]: wouldn't "?" operator be more compact here?

File: org.eclipse.ui.console/src/org/eclipse/ui/console/TextConsole.java [hunk 7]
```diff
@@ -545,19 +539,39 @@ private synchronized void checkFinished() {
...
+		var right = fPositions.size() - 1;
+
+		int mid;
+		Position position;
+
+		while (left <= right) {
+
+			mid = (left + right) / 2;
+			position = fPositions.get(mid);
+			if (position.getOffset() > offset) {
```
Comment (lbownik) [inline org.eclipse.ui.console/src/org/eclipse/ui/console/TextConsole.java]: woulnd't "?" operator be more compact here?

**Comment by laeubi** [comment]
This repository was merged with https://github.com/eclipse-platform/eclipse.platform if this PR is still relevant please rebase it and push it to the new origin!

**Pull request closed without merging**